### DISCUSSION TITLE
How to add custom CSS onto perspective-viewer

### DISCUSSION THREAD

**Discussion by mhkeller** [perspective-dev/perspective#2367] [Q&A] [answered]
I wanted to make some small CSS changes like altering the shade of green on the status circle and making column headers bold to better match the application theme. I've been reading about the shadow root but I'm definitely no expert in it. I see [this issue](https://github.com/finos/perspective/issues/1990) but that looks like it's about creating your own component.

I'm wondering what's the best pattern for injecting custom styles? Looking at the theme css files, those don't seem to cover elements like column header text styles but maybe I just wasn't able to find them. My guess is it's a bit more complicated since editing styles via dev tools has no effect.

**Comment by texodus** [comment, marked as answer]
The datagrid plugin is just a regular HTML `<table>` in the light DOM (virtualized via [`regular-table`](https://github.com/finos/regular-table)), you can style it with regular CSS in the document:

```
perspective-viewer table thead th {
    background-color: red;
}
```

You may need to be aggressive with specificity to override the rules provided by default.

Perspective adds several of metadata classes to the table elements during it's render phase as well, but these are not documented. [The `perspective-viewer-datagrd` package CSS source](https://github.com/finos/perspective/blob/master/packages/perspective-viewer-datagrid/src/less/pro.less) has the default styles and some examples of things like e.g. column type matching.

**Reply by mhkeller** [reply]
Thanks – it looks like it was a combination of factors. 1) the bold weight for that font wasn't being imported so `font-weight: bold;` had no effect 2) possibly having the right css selectors for some 3) for the status color, I get a flash of the color I want but then I'm guessing it's being overwritten via JavaScript somewhere. I'll keep playing around with it and maybe abandon some but this is very useful to know there isn't some other CSS thing in the way.

**Reply by mhkeller** [reply]
Short update to this: I was able to do a few of the key changes I wanted to make but it looks like there are some styles that exist only under the `:host` selector. It seems by design that styles in the parent aren't able to override these. It's not super high-priority for me at the moment but leaving this here in case someone else runs into this and has a solution.

**Reply by texodus** [reply]
I really need to know specifically what you're trying to override, perhaps share a repo?

The `<table>` element's `:host` is `document`, so you can override anything defined here, though TMK there is nothing in the light DOM CSS qualified by `:host`.

**Reply by mhkeller** [reply]
I'll can make a repo this evening if still needed but here's a screencap from the website as quick reference. I was looking into modifying the bottom border on the `#status_bar` element. Here's a screenshot from the website showing that it's getting that line from the `box-shadow` style on `:host #status_bar`.


https://github.com/finos/perspective/assets/498744/ac8076fb-25eb-435c-8b0a-e273e03bf38d

**Reply by texodus** [reply]
This is not part of the Datagrid, this is the `<perspective-viewer>` element, which you can only style via the exposed theme variables.

**Reply by mhkeller** [reply]
Got it. So the only way would be to change `--inactive-border-color` but that would have other effects elsewhere. Thanks for the extra info!

**Reply by texodus** [reply]
Well, it depends on what you mean by "elsewhere".  You may still be able to get away with something like this e.g. for each plugin:

```css
perspective-viewer {
--inactive--border-color: transparent;
}

perspective-viewer-datagrid {
--inactive--border-color: #FF0000; // whatever other color
}
```

.. but this is an implementation detail, we likely do (and will in the future) use this variable in a variety of other places where it is non-separable. A repo would help me understand overall what you're looking for, or better a mockup if you're willing to share. 

We've kept this API minimal so far. On `<perspective-viewer>`'s interface (the part which can be made invisible by collapsing the tab), you can override the primary and secondary colors, fonts, and icons (though using icons of different dimensions than the builtins will undoubtedly cause other issues).

**Comment by limx0** [comment]
Somewhat related (and please forgive my JS/CSS ignorance), but is it possible to hide certain tools from the status bar? Suppose I wanted to disable the `export` and `copy` buttons?

**Reply by texodus** [reply]
@limx0 No, Perspective will never support this. It is completely futile to try to prevent users from copying or downloading data that the browser itself must download anyway.

**Comment by Mindbooom** [comment]
I would like to modify the cell’s style based on perspective, such as adding up and down arrows inside the cell to change the data. Is there any way to achieve this? Thank you very much.
